File: back.js
var cellWidth     = 500;
var cellHeight    = 500;
var wholeWidth    = 2000;
var wholeHeight   = 2000;
var playerMaxNum  = 20;
var ballMaxNum    = 320;//at most 16 ball for per player
var fruitMaxNum   = 200;
var maxDist       = 100;
var playerAvailIdArray  = new Array(playerMaxNum);
var playerLiveArray     = new Array(playerMaxNum);
var playerPosXArray     = new Array(playerMaxNum);
var playerPosYArray     = new Array(playerMaxNum);
var playerDirXArray     = new Array(playerMaxNum);
var playerDirYArray     = new Array(playerMaxNum);
var ballPosXArray       = new Array(playerMaxNum);
var ballPosYArray       = new Array(playerMaxNum);
var fruitLiveArray      = new Array(fruitMaxNum);
var fruitPosXArray      = new Array(fruitMaxNum);
var fruitPosYArray      = new Array(fruitMaxNum);
var playerRadiusArray   = new Array(playerMaxNum);
var playerMaxSpeedArray = new Array(playerMaxNum);

(function(){
	for(var i = 0; i < playerMaxNum; i++)
	{
		playerAvailIdArray [i] = i;
		playerLiveArray    [i] = true;
		playerPosXArray    [i] = Math.random() * wholeWidth;
		playerPosYArray    [i] = Math.random() * wholeHeight;
		playerDirXArray    [i] = Math.random();
		playerDirYArray    [i] = Math.random();
		playerRadiusArray  [i] = 20;
		playerMaxSpeedArray[i] = 3.0;
	}
	for(var i = 0; i < fruitMaxNum; i++)
	{
		fruitLiveArray     [i] = true;
		fruitPosXArray     [i] = Math.random() * wholeWidth;
		fruitPosYArray     [i] = Math.random() * wholeHeight;
	}
})();

function getFreePlayerId()
{
	if(playerAvailIdArray.length > 0)
		return playerAvailIdArray.pop();
	else
		return -1;
}

function responseInit(obj)
{
	var playerId;
	var retObj = {};
	playerId = retObj.playerId = getFreePlayerId();
	retObj.playerRadius = playerRadiusArray[playerId];
	retObj.pos = {};
	retObj.pos.x = playerPosXArray[playerId];
	retObj.pos.y = playerPosYArray[playerId];
	return retObj;
}

function checkCollision(playerId)
{
	var playerX = playerPosXArray[playerId];
	var playerY = playerPosYArray[playerId];
	var playerR = playerRadiusArray[playerId];
	for(var i = 0; i < playerMaxNum; i++)
	{
		if(i == playerId || !playerLiveArray[i])
			continue;
		var x = playerPosXArray[i];
		var y = playerPosYArray[i];
		var r = playerRadiusArray[i];
		var dx = x-playerX;
		var dy = y-playerY;
		var dist = Math.sqrt(dx*dx+dy*dy);
		if(r + 1 < playerR && dist + r < playerR + 1)
		{
			playerR = Math.sqrt(r*r+playerR*playerR);
			playerLiveArray[i] = false;
		}
	}

	for(var i = 0; i < fruitMaxNum; i++)
	{
		if(!fruitLiveArray[i])
			continue;
		var x = fruitPosXArray[i];
		var y = fruitPosYArray[i];
		var r = 5;
		var dx = x-playerX;
		var dy = y-playerY;
		var dist = Math.sqrt(dx*dx+dy*dy);
		if(r + 1 < playerR && dist + r < playerR)
		{
			playerR = Math.sqrt(r*r+playerR*playerR + 1);
			fruitLiveArray[i] = false;
		}
	}
	playerRadiusArray[playerId] = playerR;
}

function responseUpdate(obj)
{
	/* {"header":"update", "body":{}}
	 * update-request {
	 *		"playerId":12,
	 *		"dirX":123,
	 *		"dirY":456,
	 *		"actualWidth":1366,
	 *		"actualHeight":768,
	 *		"zoomRate":1.0	}
	 * update-response {
	 *		"player":{"x":465, "y":789},
	 *		"enemy":{"x":465, "y":789},
	 *		"fruit":{"x":465, "y":789}
	 * }
	 */
	var retJSON = {};
	var playerId = obj.body.playerId;
	var actualWidth = obj.body.actualWidth;
	var actualHeight = obj.body.actualHeight;
	var playerX = playerPosXArray[playerId];
	var playerY = playerPosYArray[playerId];
	var playerR = playerRadiusArray[playerId];

	var dx = obj.body.dirX;
	var dy = obj.body.dirY;
	var dist = Math.sqrt(dx*dx+dy*dy);
	var rate = (dist > maxDist) ? 1 : (dist/maxDist);
	var playerSpeed = rate * playerMaxSpeedArray[playerId];

	var finalX = playerX + playerSpeed * dx / dist;
	var finalY = playerY + playerSpeed * dy / dist;

	retJSON.player = {};
	retJSON.enemy  = [];
	retJSON.fruit  = [];
	checkCollision(playerId);
	for(var i = 0; i < playerMaxNum; i++)
	{
		var x = playerPosXArray[i];
		var y = playerPosYArray[i];
		var r = playerRadiusArray[i];
		if(playerLiveArray[i]
		&& 2*Math.abs(x-playerX) < actualWidth
		&& 2*Math.abs(y-playerY) < actualHeight)
		{
			retJSON.enemy.push({"x":x,"y":y,"r":r});
		}
	}

	for(var i = 0; i < fruitMaxNum; i++)
	{
		var x = fruitPosXArray[i];
		var y = fruitPosYArray[i];
		if(fruitLiveArray[i]
		&& 2*Math.abs(x-playerX) < actualWidth
		&& 2*Math.abs(y-playerY) < actualHeight)
			retJSON.fruit.push({"x":x,"y":y});
	}

	if(finalX > 0 && finalX < wholeWidth)
		playerX = playerPosXArray[playerId] = finalX;
	if(finalY > 0 && finalY < wholeHeight)
		playerY = playerPosYArray[playerId] = finalY;

	retJSON.player = {"x":playerX, "y":playerY, "r":playerR};
	return retJSON;
}

/*send and post*/
function postDataToServer(dataFromClient)
{
	var retObj = {};
	var obj = JSON.parse(dataFromClient);
	switch(obj.header) {
		case 'init':
			retObj = responseInit(obj);
			break;
		case 'update':
			retObj = responseUpdate(obj);
			break;
	}
	return JSON.stringify(retObj);
}

File: back.py
#!/usr/bin/python
#-*- coding=utf-8 -*-

import log
import websocket
import math, json, random

cellWidth     = 500
cellHeight    = 500
wholeWidth    = 2000
wholeHeight   = 2000
playerMaxNum  = 20
ballMaxNum    = 320#at most 16 ball for per player
fruitMaxNum   = 200
maxDist       = 100
playerActiveIdArray = []
playerAvailIdArray  = [0] * playerMaxNum
playerWebSocketArray= [0] * playerMaxNum
playerNameArray     = [''] * playerMaxNum
playerLiveArray     = [0] * playerMaxNum
playerPosXArray     = [0.] * playerMaxNum
playerPosYArray     = [0.] * playerMaxNum
playerDirXArray     = [0.] * playerMaxNum
playerDirYArray     = [0.] * playerMaxNum
ballPosXArray       = [0.] * playerMaxNum
ballPosYArray       = [0.] * playerMaxNum
fruitLiveArray      = [0] * fruitMaxNum
fruitPosXArray      = [0.] * fruitMaxNum
fruitPosYArray      = [0.] * fruitMaxNum
playerRadiusArray   = [0.] * playerMaxNum
playerMaxSpeedArray = [0.] * playerMaxNum

def init():
    for i in range(playerMaxNum):
        playerAvailIdArray [i] = i
        playerLiveArray    [i] = True
        playerPosXArray    [i] = random.uniform(0, wholeWidth)
        playerPosYArray    [i] = random.uniform(0, wholeHeight)
        playerDirXArray    [i] = random.random()
        playerDirYArray    [i] = random.random()
        playerRadiusArray  [i] = 20
        playerMaxSpeedArray[i] = 3.0
    for i in range(fruitMaxNum):
        fruitLiveArray     [i] = True
        fruitPosXArray     [i] = random.random() * wholeWidth
        fruitPosYArray     [i] = random.random() * wholeHeight

def getFreePlayerId():
    if len(playerAvailIdArray) > 0:
        availId = playerAvailIdArray.pop()
        playerActiveIdArray.append(availId)
        return availId
    else:
        return -1

def checkCollision(playerId):
    playerX = playerPosXArray[playerId]
    playerY = playerPosYArray[playerId]
    playerR = playerRadiusArray[playerId]
    for i in playerActiveIdArray:
        if i == playerId or not playerLiveArray[i]:
            continue
        x = playerPosXArray[i]
        y = playerPosYArray[i]
        r = playerRadiusArray[i]
        dx = x-playerX
        dy = y-playerY
        dist = math.sqrt(dx*dx+dy*dy)
        if r + 1 < playerR and dist + r < playerR + 1:
            playerR = math.sqrt(r*r+playerR*playerR)
            playerLiveArray[i] = False
            playerWebSocketArray[i].send('{"header":"status", "status":"dead"}')

    for i in range(fruitMaxNum):
        if not fruitLiveArray[i]:
            continue
        x = fruitPosXArray[i]
        y = fruitPosYArray[i]
        r = 5
        dx = x-playerX
        dy = y-playerY
        dist = math.sqrt(dx*dx+dy*dy)
        if r + 1 < playerR and dist + r < playerR:
            playerR = math.sqrt(r*r+playerR*playerR + 1)
            fruitLiveArray[i] = False
    playerRadiusArray[playerId] = playerR

def responseUpdate(obj):
    retJSON = {}
    playerId = int(obj['playerId'])
    width = obj['width']
    height = obj['height']
    playerX = playerPosXArray[playerId]
    playerY = playerPosYArray[playerId]
    playerR = playerRadiusArray[playerId]

    if not playerId in playerActiveIdArray:
        return {"header":"status", "status":"dead"}

    retJSON['player'] = {}
    retJSON['enemy']  = []
    retJSON['fruit']  = []
    for i in playerActiveIdArray:
        x = playerPosXArray[i]
        y = playerPosYArray[i]
        r = playerRadiusArray[i]
        if playerLiveArray[i] \
	and 2*math.fabs(x-playerX) < width \
        and 2*math.fabs(y-playerY) < height:
            retJSON['enemy'].append({"x":x,"y":y,"r":r})

    for i in range(fruitMaxNum):
        x = fruitPosXArray[i]
        y = fruitPosYArray[i]
        if fruitLiveArray[i] \
        and 2*math.fabs(x-playerX) < width \
        and 2*math.fabs(y-playerY) < height:
            retJSON['fruit'].append({"x":x,"y":y})

    if playerLiveArray[playerId]:
        dx = obj['dirX']
        dy = obj['dirY']
        dist = math.sqrt(dx*dx+dy*dy)
        rate = [dist/maxDist, 1][dist > maxDist]
        playerSpeed = rate * playerMaxSpeedArray[playerId]
        
        finalX = playerX + playerSpeed * dx / dist
        finalY = playerY + playerSpeed * dy / dist

        if finalX > 0 and finalX < wholeWidth:
            playerX = playerPosXArray[playerId] = finalX
        if finalY > 0 and finalY < wholeHeight:
            playerY = playerPosYArray[playerId] = finalY
        
        checkCollision(playerId)

    retJSON['player'] = {"x":playerX, "y":playerY, "r":playerR}
    return retJSON

def responseInit(obj):
    retObj = {}
    playerId = int(obj['playerId'])
    if playerId == -1:
        playerId = int(getFreePlayerId())
    else:
        playerLiveArray[playerId] = True
    playerNameArray[playerId] = obj['name']
    playerPosXArray[playerId] = random.uniform(0, wholeWidth)
    playerPosYArray[playerId] = random.uniform(0, wholeHeight)
    retObj['playerId'] = playerId
    retObj['playerRadius'] = playerRadiusArray[playerId]
    retObj['pos'] = {}
    retObj['pos']['x'] = playerPosXArray[playerId]
    retObj['pos']['y'] = playerPosYArray[playerId]
    return retObj

# send and post
def handle_data(dataFromClient):
    data_valid = True
    try:
        obj = json.loads(dataFromClient)
    except:
        data_valid = False
    if not data_valid or not obj.get('header'):
        log.e('ws get invalid data (len={}):{}'.format(len(dataFromClient), ','.join('%x'%(ord(i)) for i in dataFromClient)))
        return {"header":"unknown"}
    switcher = {
            'init':responseInit,
            'update':responseUpdate,
            }
    retObj = switcher[obj['header']](obj)
    if not 'header' in retObj:
        retObj['header'] = obj['header']
    return retObj

def handler(ucon):
    playerId = -1
    ws = websocket.load(ucon)
    while 1:
        ws.recv()
        if ws.closed:
            ucon.alive = False
            break
        obj = handle_data(ws.data)
        ws.send(json.dumps(obj))
        if obj.get('header') == 'init':
            playerId = int(obj['playerId'])
            playerWebSocketArray[playerId] = ws

init()
